src/backend/optimizer/path/allpaths.c

```diff
@@ -2110,15 +2110,43 @@ set_cte_pathlist(PlannerInfo *root, RelOptInfo *rel, RangeTblEntry *rte)
 			subplan = subquery_planner(cteroot->glob, subquery, cteroot, cte->cterecursive,
 									   tuple_fraction, &subroot, config);
 
-			cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+			/*
+			 * Sharing General and SegmentGeneral subplan may lead to deadlock
+			 * when executed with 1-gang and joined with n-gang.
+			 */
+			if (CdbPathLocus_IsGeneral(*subplan->flow) ||
```

Comment:
Why did you decide to reverse the conditions? These are now contrasted with the conditions below.

Reply:
> make 1 logical operations instead of 3

it is depend on compiler

Reply:
Make 1 logical operations instead of 3 _in code_

Reply:
But the condition below with else-if was more beautiful.